server/src/main/java/keywhiz/service/permissions/PermissionCheck.java

```diff
@@ -3,9 +3,7 @@
 public interface PermissionCheck {
   boolean isAllowed(Object source, String action, Object target);
 
-  default boolean isAllowed(Object source, String action) {
-    return isAllowed(source, action, null);
-  }
+  boolean isAllowedForTargetType(Object source, String action, Class<?> targetType);
 
   default void checkAllowedOrThrow(Object source, String action, Object target){
```

Comment:
This is the crux of the change.

For our Secret creation permission check (in a future PR), we need an overload which takes in the type (class literal) of the target as a third argument. Unfortunately, this means that our existing isAllowed(Object, String, Object) signature is a superset (because Object is a superclass of Class) of our new overload, which makes it hard to reason about dispatch and also confuses Mockito, leading to unexpected behavior and easy-to-introduce bugs.

So, as a solution, I decided to change the method name of the overload from isAllowed to isAllowedForTargetType to eliminate any chance of conflict. I also decided to get rid of the default implementations of this method and the original isAllowed as a hint to implementers that they needed to reason about both of these cases separately.

Finally, with our new overload, the existing two-argument isAllowed(Object, String) isn't used anywhere else, so I removed it. In hindsight, it really didn't make much sense to have it automatically delegate to isAllowed(Object, String, Object) since those are separate questions, and because it could also just as easily delegate to isAllowedForTargetType(Object, String, Class).